audio/quarto_mexca_performance_diarization.qmd

```diff
@@ -0,0 +1,378 @@
+    
+    reference_speakers <- unique(speaker_ref_segments$name)
+    confusion_matrix <- as.matrix(table(levels(reference_speakers), 
+                                        levels(reference_speakers)))
+    
+    for (ref_speaker1 in reference_speakers){
+      sub_speaker_ref1 <- speaker_ref_segments[speaker_ref_segments$name == ref_speaker1,]
+      
+      sub_speaker_pred1 <- speaker_pred_segments[speaker_pred_segments$mapped_reference == ref_speaker1,]
+      
+      for(ref_speaker2 in reference_speakers){
+        sub_speaker_ref2 <- speaker_ref_segments[speaker_ref_segments$name == ref_speaker2,]
+        
+        sub_speaker_pred2 <- speaker_pred_segments[speaker_pred_segments$mapped_reference == ref_speaker2,]
+        
+        # compute speech segments overlap (in seconds) between mexca and the reference
+        confusion_matrix[ref_speaker1,ref_speaker2] <- calculate_overlap(sub_speaker_ref1, sub_speaker_pred2)
+        
+      }
+      
+      # store matrix 
+      my_confusion_2d[[i,j]] <- confusion_matrix
+      
+      # store plot 
+      my2Dlist[[i,j]] <- confusion_matrix|>
+        as.data.frame()|>
+        ggplot(ggplot2::aes(Var1, Var2, fill = Freq)) +
+        geom_tile(
+          colour = "white",
+          lwd = 0.1,
+          linetype = 1) +
+        geom_text(aes(label = round(Freq,1))) +
+        scale_fill_gradient(low="white", high="#009194")+
+        coord_fixed()+
+        guides(fill = guide_colourbar(title = "seconds"))+
+        ylab("True speaker label")+
+        xlab("Predicted speaker label")+
+        labs(title = paste0("n = ", n, " max duration = ", max(confusion_matrix), "s" ),
+             caption="Dutch debate dataset 2021")+
+        theme(legend.position="right")
+      
+      
+    }
+  }
+}
+
+
+```
+
+At this point we're ready to compute the f1 score for each speaker, in each dataset, and store it in the `F1_total` dataframe.
+
+```{r}
+F1_total <- data.frame()
+
+for(i in 1:nrow(my_confusion_2d)){ #n of speakers
+  for(j in 1:ncol(my_confusion_2d)){ # max_duration
+    
+    n <- n_speaker[i]
+    t <- max_duration[j]
+    
+    speaker_ref_segments <- droplevels(reference_ded21_for_pyannote[reference_ded21_for_pyannote$n_speaker == n & reference_ded21_for_pyannote$max_duration == t,])
+    
+    confusion_matrix <- my_confusion_2d[[i,j]]
+    reference_speakers <- colnames(confusion_matrix)
+    
+    F1 = data.frame(
+      speakers = reference_speakers,
+      max_duration = NA,
+      f1_score = NA,
+      n_speakers = n,
+      file = t
+      
+    )
+    
+    for (k in reference_speakers){
+        F1[F1$speakers==k,]$f1_score <- compute_f1(confusion_matrix, k)
+    
+        # append speakers' individual max duration
+        F1[F1$speakers==k,]$max_duration <- sum(speaker_ref_segments[speaker_ref_segments$name == k,]$tdur)
+
+    }
+    
+    F1_total <- bind_rows(F1_total, F1)
+
+  }
+}
+```
+
+```{r}
+head(F1_total)
+```
+
+## F1 score distribution
```

Comment:
Maybe I missed this: But is the following analysis about the balanced or unbalanced data set (or both) and is there a difference between the two data sets?

Reply:
Thanks for noticing this! this analysis is on the unbalanced dataset, as I thought this would be more informative. I have specified that now in the quarto notebook